### PR TITLE
Doc(eos_designs): add precisions regarding eos_designs and eos_cli_config_gen variables

### PR DESCRIPTION
**Pull request by gmuloc**
## Change Summary

This PR intends to add a section in the documentation with the current behaviour of the 

## Related Issue(s)

None

## Component(s) name

`arista.avd.eos_designs`

## Proposed changes

Only documentation change describing the behaviours (with examples) and suggesting some recommendations.

## How to test

no test

### User Checklist

- N/A

### Repository Checklist

- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [x] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

### CODE REVIEW

File: ansible_collections/arista/avd/roles/eos_designs/doc/custom-structured-configuration.md [hunk 1]
```diff
@@ -147,7 +146,7 @@ my_dci_ethernet_interfaces:
     peer_interface: Ethernet123
     peer_type: my_precious
 
-my_special_dci_ethernet_interfaces:
+
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/doc/custom-structured-configuration.md]: remove extra empty line

File: ansible_collections/arista/avd/roles/eos_designs/doc/custom-structured-configuration.md [hunk 2]
```diff
@@ -191,3 +190,157 @@ name_server:
   - 10.10.10.11
   - 10.10.10.12
 ```
+## Variables Precedence
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/doc/custom-structured-configuration.md]: I think this description should be moved up above the structured config data model stuff. So the user understands the problem before trying to solve it.

File: ansible_collections/arista/avd/roles/eos_designs/doc/custom-structured-configuration.md [hunk 3]
```diff
@@ -191,3 +190,157 @@ name_server:
...
+
+_**TL;DR:** eos_cli_config_gen variables in the input files will be overwritten
+by the eos_designs generated facts when the role produces a value for the same
+key, use custom_strutcured_configuration to make sure the values make it in the intended configurations._
+
+### Overview
+
+AVD `eos_designs` role flow is expected to work as follow:
+
+1. from the input variables (inventory, group_vars, host_vars,...), a task first sets ansible facts for each device
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/doc/custom-structured-configuration.md]: For this description I think the flow could be simplified to:

1. from the input variables (inventory, group_vars, host_vars, ...) the `eos_designs` role generates the "structured_config"
2. The "structured_config" is used as input to the `eos_cli_config_gen` role to generate the intended EOS configuration.

File: ansible_collections/arista/avd/roles/eos_designs/doc/custom-structured-configuration.md [hunk 4]
```diff
@@ -191,3 +190,157 @@ name_server:
...
+2. the facts are then used in a subsequent task to generate a structured_config
+3. this structured_config is what is then used by `eos_cli_config_gen` role to generate the intended config
+
+A caveat is that it is possible to define in the inventory, group_vars, host_vars,... some variables that are
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/doc/custom-structured-configuration.md]:
```suggestion
A caveat is that in the inventory, group_vars, host_vars,... you can define some variables that are
```

File: ansible_collections/arista/avd/roles/eos_designs/doc/custom-structured-configuration.md [hunk 5]
```diff
@@ -191,3 +190,157 @@ name_server:
...
+used by `eos_designs` and others used by `eos_cli_config_gen` roles. And while it is perfectly acceptable,
+the behavior can sometimes be confusing. If the key name for an `eos_cli_config_gen` variable in the variable files is the same as one of
+the key name of a variable generated by `eos_designs` steps, only the value from `eos_designs` step will be configured.
+However if there is no clash, then the value of the variable will be the one defined in the variable files.
+
+The consequence is that upgrading AVD to a newer version for instance, or configuring additional `eos_designs` functionnality in some existing AVD repo,
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/doc/custom-structured-configuration.md]:
```suggestion
The consequence is that upgrading AVD to a newer version for instance, or configuring additional `eos_designs` functionnality in an existing AVD project,
```

File: ansible_collections/arista/avd/roles/eos_designs/doc/custom-structured-configuration.md [hunk 6]
```diff
@@ -191,3 +190,157 @@ name_server:
...
+could lead to some additional variable keys being generated by `eos_designs` in the structured_config, overwriting previously working `eos_cli_config_gen` keys.
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/doc/custom-structured-configuration.md]:
```suggestion
could lead to additional variable keys being generated by `eos_designs` in the structured_config, overwriting previously working `eos_cli_config_gen` keys.
```

File: ansible_collections/arista/avd/roles/eos_designs/doc/custom-structured-configuration.md [hunk 7]
```diff
@@ -191,3 +190,157 @@ name_server:
...
+
+<!-- ![Figure 1: Variables Precedence between roles](../../../media/variables_precedence.png) -->
+
+<div style="text-align:center">
+  <img src="../../../media/variables_precedence.png" />
+</div>
+
+The recommendation is to use custom_structured_configuration as described in [the following section of this guide](./custom-structured-configuration.md). This configuration will always prevails over the `eos_designs` generated variables.
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/doc/custom-structured-configuration.md]: After you move this up the vars for custom_structured_configuration will be below on this page.
```suggestion
The recommendation is to use `custom_structured_configuration` as described below. This configuration will be merged on top of the `eos_designs` generated variables.
```

File: ansible_collections/arista/avd/roles/eos_designs/doc/custom-structured-configuration.md (hunk 1 above)
Comment (gmuloc) [reply ansible_collections/arista/avd/roles/eos_designs/doc/custom-structured-configuration.md]: actually this line (and the one above) should not have been removed.. fixing this

File: ansible_collections/arista/avd/roles/eos_designs/doc/custom-structured-configuration.md (hunk 2 above)
Comment (gmuloc) [reply ansible_collections/arista/avd/roles/eos_designs/doc/custom-structured-configuration.md]: moved on top

File: ansible_collections/arista/avd/roles/eos_designs/doc/custom-structured-configuration.md (hunk 3 above)
Comment (gmuloc) [reply ansible_collections/arista/avd/roles/eos_designs/doc/custom-structured-configuration.md]: addressed in [215e84d](https://github.com/aristanetworks/ansible-avd/pull/1743/commits/215e84d71a018127cdd37073dde4c374dc962906)

File: ansible_collections/arista/avd/roles/eos_designs/doc/custom-structured-configuration.md [hunk 8]
```diff
@@ -3,6 +3,160 @@
...
+1. from the input variables (inventory, group_vars, host_vars, ...) the `eos_designs` role generates the "structured_config"
+2. The "structured_config" is used as input to the `eos_cli_config_gen` role to generate the intended EOS configuration.
+
+A caveat is that in the inventory, group_vars, host_vars,... you can define some variables that are
+used by `eos_designs` and others that are used by `eos_cli_config_gen` roles. And while it is perfectly acceptable,
+the behavior can sometimes be confusing. If the key name for an `eos_cli_config_gen` variable in the variable files is the same as one of
+the key name of a variable generated by `eos_designs` steps, only the value from `eos_designs` step will be configured.
+However if there is no clash, then the value of the variable will be the one defined in the variable files.
+
+The consequence is that upgrading AVD to a newer version for instance, or configuring additional `eos_designs` functionnality in an existing AVD project,
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/doc/custom-structured-configuration.md]:
```suggestion
The consequence is that upgrading AVD to a newer version for instance, or configuring additional `eos_designs` functionality in an existing AVD project,
```

**Review by ClausHolbechArista — APPROVED** [review]
Single typo. Other than that LGTM

File: ansible_collections/arista/avd/roles/eos_designs/doc/custom-structured-configuration.md [hunk 9]
```diff
@@ -3,6 +3,160 @@
...
+
+```
+ip name-server vrf EOS_CLI 4.4.4.4
+ip name-server vrf EOS_CLI 8.8.8.8
+```
+
+#### eos_cli_config_gen variables overwritten by eos_designs variables
+
+```
+# Both name_servers from eos_design and name_server from
```
Comment (emilarista) [inline ansible_collections/arista/avd/roles/eos_designs/doc/custom-structured-configuration.md]: Typo: eos_design

File: ansible_collections/arista/avd/roles/eos_designs/doc/custom-structured-configuration.md [hunk 10]
```diff
@@ -3,6 +3,160 @@
...
+ip name-server vrf MGMT 192.168.42.10
+ip name-server vrf MGMT 192.168.42.40
+```
+
+
+#### eos_designs variables overwritten by custom_structured_configuration
+
+```
+---
+# Both name_servers from eos_design and leveraging the
```
Comment (emilarista) [inline ansible_collections/arista/avd/roles/eos_designs/doc/custom-structured-configuration.md]: Same typo.

**Review by emilarista — APPROVED** [review]
Small typos, otherwise LGTM!

**Pull request merged**